### PR TITLE
program information serialization

### LINKED ISSUE #82: Program information returned from extensions should be strongly typed
@shankarseal

### PR DESCRIPTION
**Pull request by shankarseal**
This change fixes #82 and #289 .

The summary of the changes:
1) ebpf_extension_data_t.data is now an opaque pointer instead of byte buffer. Program information NPI provider now sets the opaque pointer to ebpf_program_data_t which in turn has pointers to program_information and array of helper function addresses.
2) The helper function addresses are no longer passed in as program informatoin NPI provider's dispatch table, as these functions do not constitute the NPI contract between NPI modules (EC and extension). Rather they are the contract between eBPF program and Extension.
3). The program_information (includes program type context and helper function prototypes) are serialized in the GET_PROGRAM_INFORMATION protocol message handler and de-serialized in get_program_information_data.
4) Changes have been made in ebpf_program and trampoline modules to read the addresses from the provider data.

On a separate PR I will add a test extension.
Further changes will be made to the NMR code in both EC and net_ebpf_ext as part of issues #291 .

### CODE REVIEW

File: tests/end_to_end/helpers.h [hunk 1]
```diff
@@ -99,103 +99,81 @@ typedef class _single_instance_hook
...
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data_meta)};
+static ebpf_program_information_t _ebpf_xdp_program_information = {{"xdp",
+                                                                    &_ebpf_xdp_context_descriptor,
+                                                                    // clang-format off
+                                                                    {// EBPF_PROGRAM_TYPE_XDP
+                                                                        0xf1832a85,
+                                                                        0x85d5,
+                                                                        0x45b0,
+                                                                        0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0}},
+                                                                     // clang-format om
```
Comment (Alan-Jowett) [inline tests/end_to_end/helpers.h:128]: Typo - Should be on, not om.

**Comment by Alan-Jowett** [comment]
Remove dead code? https://github.com/microsoft/ebpf-for-windows/tree/master/tools/encode_program_information is all dead code now.

File: libs/platform/ebpf_serialize.h [hunk 2]
```diff
@@ -82,7 +82,55 @@ extern "C"
...
     void
-    ebpf_map_information_array_free(uint16_t map_count, _In_count_(map_count) ebpf_map_information_t* map_info);
+    ebpf_map_information_array_free(
+        uint16_t map_count, _Pre_maybenull_ _Post_invalid_ _In_count_(map_count) ebpf_map_information_t* map_info);
+
+    /**
+     * @brief Serialize ebpf_program_information_t onto output buffer.
+     *
+     * @param[in]  program_info Pointer to program_map_information_t to serialize.
+     * @param[out]  output_buffer Caller specified output buffer to write serialized data into.
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.h:92]:
```suggestion
     * @param[out] output_buffer Caller specified output buffer to write serialized data into.
```

File: libs/platform/ebpf_serialize.h [hunk 3]
```diff
@@ -82,7 +82,55 @@ extern "C"
...
+        _Outptr_ ebpf_program_information_t** program_info);
+
+    /**
+     * @brief Helper Function to free ebpf_program_information_t allocated by
+     * ebpf_deserialize_program_information function.
+     *
+     * @param[in] program_info Program information to be freed.
+     */
+    void
+    ebpf_program_information_free(_Pre_maybenull_ _Post_invalid_ _In_ ebpf_program_information_t* program_info);
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.h:133]: Can you just use the shorter
`_In_opt_ _Post_invalid_`
or is the meaning different between `_opt_` vs `_Pre_maybenull_`?

File: netebpfext/net_ebpf_ext.c [hunk 4]
```diff
@@ -48,11 +48,64 @@ static ebpf_ext_attach_hook_provider_registration_t* _ebpf_xdp_hook_provider_reg
...
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data),
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data_end),
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data_meta)};
+static ebpf_program_information_t _ebpf_xdp_program_information = {{"xdp",
+                                                                    &_ebpf_xdp_context_descriptor,
+                                                                    {// EBPF_PROGRAM_TYPE_XDP
+                                                                     0xf1832a85,
+                                                                     0x85d5,
+                                                                     0x45b0,
+                                                                     // clang-format off
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:78]: why is clang-format off needed here?

File: netebpfext/net_ebpf_ext.c [hunk 5]
```diff
@@ -48,11 +48,64 @@ static ebpf_ext_attach_hook_provider_registration_t* _ebpf_xdp_hook_provider_reg
...
+                                                                     0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0}},
+                                                                   // clang-format on
+                                                                   EBPF_COUNT_OF(_ebpf_map_helper_function_prototype),
+                                                                   _ebpf_map_helper_function_prototype};
+
+static ebpf_program_data_t _ebpf_xdp_program_data = {&_ebpf_xdp_program_information, NULL};
+
+static ebpf_extension_data_t _ebpf_xdp_program_information_provider_data = {0, // version
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:86]: use a version define instead of hard coding 0 with a comment

File: tests/api_test/api_test.vcxproj [hunk 6]
```diff
@@ -82,7 +82,7 @@
     <LinkIncremental>false</LinkIncremental>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
-    <LinkIncremental>true</LinkIncremental>
+    <LinkIncremental>false</LinkIncremental>
```
Comment (dthaler) [inline tests/api_test/api_test.vcxproj:85]: What do the changes in this file do?  I'm trying to match them with something in the PR description...
Same question on the other vcxproj files in this PR

File: tests/end_to_end/helpers.h [hunk 7]
```diff
@@ -99,103 +99,81 @@ typedef class _single_instance_hook
...
+     EBPF_RETURN_TYPE_INTEGER,
+     {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY}}};
+
+static ebpf_context_descriptor_t _ebpf_xdp_context_descriptor = {sizeof(xdp_md_t),
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data),
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data_end),
+                                                                 EBPF_OFFSET_OF(xdp_md_t, data_meta)};
+static ebpf_program_information_t _ebpf_xdp_program_information = {{"xdp",
+                                                                    &_ebpf_xdp_context_descriptor,
+                                                                    // clang-format off
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:122]: why is clang-format off needed here?

File: libs/platform/unit/platform_unit_test.cpp [hunk 8]
```diff
@@ -371,3 +375,82 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+    ebpf_helper_function_prototype_t helper_prototype[] = {
+        {1000,
+         "helper_0",
+         EBPF_RETURN_TYPE_PTR_TO_MAP_VALUE_OR_NULL,
+         {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY}},
+        {1001,
+         "helper_1",
+         EBPF_RETURN_TYPE_INTEGER,
+         {EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_KEY, EBPF_ARGUMENT_TYPE_PTR_TO_MAP_VALUE}}};
+    ebpf_context_descriptor_t context_descriptor = {32, 0, 8, -1};
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:392]: Unless you're explicitly trying to enter invalid values for a case that should fail, don't hard code the size/offset numbers, use sizeof(), field offsets, etc

File: libs/platform/unit/platform_unit_test.cpp [hunk 9]
```diff
@@ -371,3 +375,82 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+    GUID PROGRAM_TYPE_TEST = {0x7ebe418c, 0x76dd, 0x4c2c, {0x99, 0xbc, 0x5c, 0x48, 0xa2, 0x30, 0x4b, 0x90}};
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:393]:
```suggestion
    GUID PROGRAM_TYPE_TEST = {0x7ebe418c, 0x76dd, 0x4c2c, {0x99, 0xbc, 0x5c, 0x48, 0xa2, 0x30, 0x4b, 0x90}};
```
```suggestion
    GUID program_type_test = {0x7ebe418c, 0x76dd, 0x4c2c, {0x99, 0xbc, 0x5c, 0x48, 0xa2, 0x30, 0x4b, 0x90}};
```
ALL_CAPS is for defines/macros/enum values

File: libs/platform/unit/platform_unit_test.cpp [hunk 10]
```diff
@@ -371,3 +375,82 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+            out_program_info->program_type_descriptor.name,
+            EBPF_MAX_PROGRAM_DESCRIPTOR_NAME_LENGTH) == 0);
+    REQUIRE(in_program_info.count_of_helpers == out_program_info->count_of_helpers);
+    REQUIRE(out_program_info->helper_prototype != nullptr);
+    for (uint32_t i = 0; i < in_program_info.count_of_helpers; i++) {
+        ebpf_helper_function_prototype_t* in_prototype = &in_program_info.helper_prototype[i];
+        ebpf_helper_function_prototype_t* out_prototype = &out_program_info->helper_prototype[i];
+        REQUIRE(in_prototype->helper_id == out_prototype->helper_id);
+        REQUIRE(in_prototype->return_type == out_prototype->return_type);
+        for (int j = 0; j < 5; j++)
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:445]:
```suggestion
        for (int j = 0; j < _countof(in_prototype->arguments); j++)
```

File: libs/platform/unit/platform_unit_test.cpp [hunk 11]
```diff
@@ -371,3 +375,82 @@ TEST_CASE("serialize_map_test", "[platform]")
...
+            REQUIRE(in_prototype->arguments[j] == out_prototype->arguments[j]);
+        REQUIRE(out_prototype->name != nullptr);
+        REQUIRE(strncmp(in_prototype->name, out_prototype->name, EBPF_MAX_HELPER_FUNCTION_NAME_LENGTH) == 0);
+    }
+
+    // Free de-serialized program information.
+    ebpf_program_information_free(out_program_info);
+
+    if (buffer != nullptr)
+        free(buffer);
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:455]:
```suggestion
    free(buffer);
```
reduce code complexity... free() is always safe to call with nullptr

File: libs/api_common/windows_helpers.cpp [hunk 12]
```diff
@@ -24,44 +26,61 @@ struct guid_compare
     }
 };
 
-static thread_local std::map<GUID, ebpf_helper::ebpf_memory_ptr, guid_compare> _program_information_cache;
+struct _ebpf_program_information_deleter
+{
+    void
+    operator()(_Pre_maybenull_ _Post_invalid_ _In_ ebpf_program_information_t* program_info)
```
Comment (dthaler) [inline libs/api_common/windows_helpers.cpp:32]: See my question elsewhere in this PR about `_opt_` vs `_Pre_maybenull_`

File: libs/api_common/windows_helpers.cpp [hunk 13]
```diff
@@ -24,44 +26,61 @@ struct guid_compare
...
+    {
+        ebpf_program_information_free(program_info);
+    }
+};
+
+typedef std::unique_ptr<ebpf_program_information_t, _ebpf_program_information_deleter> ebpf_program_information_ptr;
```
Comment (dthaler) [inline libs/api_common/windows_helpers.cpp:38]: types should end in `_t`

File: libs/platform/ebpf_serialize.c [hunk 14]
```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
 
     return result;
 }
+
+void
+ebpf_program_information_free(_Pre_maybenull_ _Post_invalid_ _In_ ebpf_program_information_t* program_info)
+{
+    if (program_info != NULL) {
+        ebpf_free(program_info->program_type_descriptor.context_descriptor);
+        ebpf_free((void*)program_info->program_type_descriptor.name);
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:223]: curious... I thought you could pass any pointer into a void* argument without complaint, why is the cast needed?

File: libs/platform/ebpf_serialize.c [hunk 15]
```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
...
+    }
+}
+
+ebpf_result_t
+ebpf_serialize_program_information(
+    _In_ const ebpf_program_information_t* program_info,
+    _Out_writes_bytes_to_(output_buffer_length, *serialized_data_length) uint8_t* output_buffer,
+    size_t output_buffer_length,
+    _Out_ size_t* serialized_data_length,
+    _Out_ size_t* required_length)
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:239]: Why do you need both of these?  Isn't serialized_data_length always 0 on failure and required_length on success?

File: libs/platform/ebpf_serialize.c [hunk 16]
```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
...
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    serialized_program_type_descriptor = (ebpf_serialized_program_type_descriptor_t*)current;
+
+    local_program_type_descriptor->platform_specific_data = serialized_program_type_descriptor->platform_specific_data;
+    local_program_type_descriptor->is_privileged = serialized_program_type_descriptor->is_privileged;
+
+    // Allocate and desrialize context_descriptor, if present.
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:446]:
```suggestion
    // Allocate and deserialize context_descriptor, if present.
```

File: libs/platform/ebpf_serialize.c [hunk 17]
```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
...
+    // Adjust remaining buffer length.
+    result = ebpf_safe_size_t_subtract(buffer_left, serialized_program_type_descriptor->name_length, &buffer_left);
+    if (result != EBPF_SUCCESS)
+        goto Exit;
+
+    // Advance the input buffer current pointer to the end of program type descriptor section.
+    current += serialized_program_type_descriptor->size;
+
+    if (buffer_left == 0) {
+        // No buffer left. This means there are no helper function prototypes information in the input buffer.
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:496]:
```suggestion
        // No buffer left. This means there are no helper function prototypes in the input buffer.
```
fix grammar

File: libs/platform/ebpf_serialize.c [hunk 18]
```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
...
+
+Exit:
+    if (result == EBPF_SUCCESS) {
+        *program_info = local_program_info;
+        local_program_info = NULL;
+    }
+
+    if (local_program_info != NULL) {
+        ebpf_program_information_free(local_program_info);
+    }
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:594]: make semantics match free() for consistency, i.e., move the NULL check inside the function

File: tests/end_to_end/helpers.h (hunk 7 above)
Comment (shankarseal) [reply tests/end_to_end/helpers.h:122]: As otherwise it was printing the guid as such:

```
    0xf1832a85,
    0x85d5,
    0x45b0,
    0x98,
    0xa0,
    0x70,
    0x69,
    0xd6,
    0x30,
    0x13,
    0xb0
```
I wanted it to show in the same format as 
uuidgen -s
INTERFACENAME = { /* 81ee4c5e-7599-4e72-b0c4-1be21b9b1841 */
    0x81ee4c5e,
    0x7599,
    0x4e72,
    {0xb0, 0xc4, 0x1b, 0xe2, 0x1b, 0x9b, 0x18, 0x41}
  };

File: libs/platform/ebpf_serialize.c (hunk 14 above)
Comment (shankarseal) [reply libs/platform/ebpf_serialize.c:227]: I got a compiler warning (treated as error) due to the cosnt qualifier in program_type_descriptor.name field.

```
F:\ebpf-for-windows\my\libs\platform\ebpf_serialize.c(225,61): warning C4090: 'function': different 'const' qu
       alifiers [F:\ebpf-for-windows\my\libs\platform\user\platform_user.vcxproj]

typedef struct _ebpf_program_type_descriptor
    {
    MIDL([string])
    const char* name;
    ebpf_context_descriptor_t *context_descriptor;
    GUID platform_specific_data;
    unsigned char is_privileged;
    } 	ebpf_program_type_descriptor_t;
```

**Comment by shankarseal** [comment]
> Remove dead code? https://github.com/microsoft/ebpf-for-windows/tree/master/tools/encode_program_information is all dead code now.

Filed #308

File: libs/platform/ebpf_serialize.c (hunk 15 above)
Comment (shankarseal) [reply libs/platform/ebpf_serialize.c:243]: The IOCTLs all use protocol request/reply messages which have a header field with length field - that shows the length of the entire message.
If the reply message is of variable length, it has an additional size field (such as in map_information and program_information reply cases). Given this, I thought it was convenient to return 2 output size parameters, specially in the failure case, as reply->size will get filled with 0, whereas required_length could be used to populate the length. Yes, in success case they both will be same.
If you see any issue with this, I can file a new github issue and address it in a separate PR.

File: libs/execution_context/ebpf_core.c [hunk 19]
```diff
@@ -699,21 +693,29 @@ _ebpf_core_protocol_get_program_information(
...
+    serialization_buffer_size = reply_length - EBPF_OFFSET_OF(ebpf_operation_get_program_information_reply_t, data);
 
-    required_length = EBPF_OFFSET_OF(ebpf_operation_get_program_information_reply_t, data);
-    required_length += program_information_data->size;
+    // Serialize program information structure onto reply data buffer.
+    retval = ebpf_serialize_program_information(
+        program_data->program_information,
+        reply->data,
+        serialization_buffer_size,
+        (size_t*)&reply->size,
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/execution_context/ebpf_core.c]: Thanks Dave. I fixed this and also a similar issue in map_information serialization struct.

File: tests/end_to_end/helpers.h [hunk 20]
```diff
@@ -99,103 +99,81 @@ typedef class _single_instance_hook
...
+                                                                        0x85d5,
+                                                                        0x45b0,
+                                                                        0x98, 0xa0, 0x70, 0x69, 0xd6, 0x30, 0x13, 0xb0}},
+                                                                   // clang-format on
+                                                                   EBPF_COUNT_OF(_ebpf_map_helper_function_prototype),
+                                                                   _ebpf_map_helper_function_prototype};
+
+static ebpf_program_data_t _ebpf_xdp_program_data = {&_ebpf_xdp_program_information, NULL};
+
+static ebpf_extension_data_t _ebpf_xdp_program_information_provider_data = {0, // version
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:134]: Place 0 with a define

File: tests/end_to_end/helpers.h [hunk 21]
```diff
@@ -99,103 +99,81 @@ typedef class _single_instance_hook
...
+                                                                        0x6c52,
+                                                                        0x4a26,
+                                                                        0xb6, 0x77, 0xbb, 0x1c, 0x34, 0x42, 0x5a, 0xdf}},
+                                                                    // clang-format on
+                                                                    EBPF_COUNT_OF(_ebpf_map_helper_function_prototype),
+                                                                    _ebpf_map_helper_function_prototype};
+
+static ebpf_program_data_t _ebpf_bind_program_data = {&_ebpf_bind_program_information, NULL};
+
+static ebpf_extension_data_t _ebpf_bind_program_information_provider_data = {0, // version
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:154]: replace 0 with a define

File: libs/api_common/windows_helpers.cpp [hunk 22]
```diff
@@ -24,44 +26,61 @@ struct guid_compare
...
 clear_program_information_cache()
 {
     _program_information_cache.clear();
 }
 
-uint32_t
-get_program_information_data(ebpf_program_type_t program_type, ebpf_extension_data_t** program_information_data)
+ebpf_result_t
+get_program_information_data(
+    ebpf_program_type_t program_type, _Outptr_result_maybenull_ ebpf_program_information_t** program_information)
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/api_common/windows_helpers.cpp:51]: How does SAL interpret return value of 0 as success? I think there has to be another annotation to instruct SAL about that?

[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/api_common/windows_helpers.cpp:51]: https://github.com/microsoft/ebpf-for-windows/blob/576ef7381cebfa1106a857c4263fb3fccda907f6/include/ebpf_result.h#L16

File: libs/execution_context/ebpf_program.c [hunk 23]
```diff
@@ -326,12 +358,13 @@ static ebpf_result_t
...
-                    EBPF_OFFSET_OF(ebpf_extension_dispatch_table_t, function)) /
-                   sizeof(program->global_helper_provider_dispatch_table->function);
+    ebpf_program_data_t* global_helper_program_data = (ebpf_program_data_t*)program->global_helper_provider_data->data;
+    ebpf_helper_function_addresses_t* global_helper_function_addresses =
+        global_helper_program_data->helper_function_addresses;
+    size_t count = global_helper_function_addresses->helper_function_count;
 
     for (index = 0; index < count; index++) {
-        const void* helper = (void*)program->global_helper_provider_dispatch_table->function[index];
+        const void* helper = (void*)global_helper_function_addresses->helper_function_address[index];
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/execution_context/ebpf_program.c:367]: Yes, as helper_function_address is an array of uint64_t

```
warning C4047: 'initializing': 'const voi
       d *' differs in levels of indirection from 'uint64_t' [F:\ebpf-for-windows\my\libs\execution_context\user\execut
       ion_context_user.vcxproj]
```

File: libs/execution_context/ebpf_program.c [hunk 24]
```diff
@@ -50,44 +51,62 @@ typedef struct _ebpf_program
...
+        }
+
+        return_value = ebpf_update_trampoline_table(program->trampoline_table, helper_function_addresses);
+        if (return_value != EBPF_SUCCESS)
+            goto Exit;
     }
 
-    program->program_information_provider_dispatch_table = provider_dispatch_table;
     program->program_information_binding_context = provider_binding_context;
     program->program_information_data = provider_data;
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_program.c:106]: Should we update the program_information_data here? It is possible extension may have changed the program context information.

File: libs/platform/ebpf_serialize.c [hunk 25]
```diff
@@ -4,25 +4,61 @@
...
+#include "ebpf_program_types.h"
+
+/**
+ * @brief Serialized program type descriptor.
+ */
+typedef struct _ebpf_serialized_program_type_descriptor
+{
+    size_t size;
+    ebpf_context_descriptor_t context_descriptor;
+    GUID platform_specific_data;
```
Comment (saxena-anurag) [inline libs/platform/ebpf_serialize.c:16]: I think we can the name "program_type" instead of "platform_specific_data" here?

"platform_specific_data" is used in the verifier code, as it can run both on Linux and Windows.

Comment (dthaler) [reply libs/platform/ebpf_serialize.c:16]: Agree, good point

File: libs/execution_context/ebpf_program.c (hunk 24 above)
Comment (shankarseal) [reply libs/execution_context/ebpf_program.c:98]: I have made some changes to ensure upon provider reload helper function count is not changed. #313 created to track other issues.

File: tests/end_to_end/helpers.h (hunk 7 above)
Comment (shankarseal) [reply tests/end_to_end/helpers.h]: Fixed.

File: libs/platform/ebpf_serialize.c [hunk 26]
```diff
@@ -174,8 +214,386 @@ ebpf_deserialize_map_information_array(
...
+
+Exit:
+    return result;
+}
+
+ebpf_result_t
+ebpf_deserialize_program_information(
+    size_t input_buffer_length,
+    _In_reads_bytes_(input_buffer_length) const uint8_t* input_buffer,
+    _Outptr_ ebpf_program_information_t** program_info)
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/platform/ebpf_serialize.c:411]: I have filed https://github.com/microsoft/ebpf-for-windows/issues/314 to track this issue.

File: libs/platform/ebpf_serialize.c [hunk 27]
```diff
@@ -174,8 +214,386 @@ ebpf_deserialize_map_information_array(
...
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_program_information_t* local_program_info;
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/platform/ebpf_serialize.c:414]: #314

File: libs/platform/unit/platform_unit_test.cpp (hunk 8 above)
Comment (dthaler) [reply libs/platform/unit/platform_unit_test.cpp:396]: Unresolving since code still has 32, 0, and 8 hard coded and I didn't see a response.  If the values are completely arbitrary and not used for anything in this test, add a // comment saying so.

File: netebpfext/net_ebpf_ext.c [hunk 28]
```diff
@@ -48,11 +48,48 @@ static ebpf_ext_attach_hook_provider_registration_t* _ebpf_xdp_hook_provider_reg
 static ebpf_ext_attach_hook_provider_registration_t* _ebpf_bind_hook_provider_registration = NULL;
 static ebpf_extension_provider_t* _ebpf_xdp_program_information_provider = NULL;
 static ebpf_extension_provider_t* _ebpf_bind_program_information_provider = NULL;
-static ebpf_extension_data_t* _ebpf_xdp_program_information_provider_data = NULL;
-static ebpf_extension_data_t* _ebpf_bind_program_information_provider_data = NULL;
 
 #define RTL_COUNT_OF(arr) (sizeof(arr) / sizeof(arr[0]))
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.c:52]: Line 42 includes ebpf_platform.h which defines EBPF_COUNT_OF, so RTL_COUNT_OF should be removed from this file.  Ok to do in a separate PR if you prefer, since it was bad before this PR.

**Review by dthaler — APPROVED** [review]
Marking Approve since I only have one remaining comment in platform_unit_test.cpp which is a nit, not a blocker.  I'd still like to see it addressed though.

File: libs/platform/unit/platform_unit_test.cpp (hunk 8 above)
Comment (shankarseal) [reply libs/platform/unit/platform_unit_test.cpp:398]: I thought I replied before resolving the thread. Not sure what happened. Yes, these are arbitrary values and have no consequence to the serialization process. Left a comment like you suggested.

File: netebpfext/net_ebpf_ext.c (hunk 28 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.c:52]: Will do so in a separate PR.

File: libs/api_common/windows_helpers.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (shankarseal) [reply libs/api_common/windows_helpers.cpp]: Thanks Alan. I have changed the SAL to _Outptr_.
